ext/drpttracer/drpttracer.h

```diff
@@ -0,0 +1,238 @@
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+/* DynamoRIO Intel PT Tracing Extension. */
+
+#ifndef _DRPTTRACER_H_
+#define _DRPTTRACER_H_ 1
+
+/**
+ * @file drpttracer.h
+ * @brief Header for DynamoRIO Intel PT Tracing Extension
+ */
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include <inttypes.h>
+
+#ifndef IN
+#    define IN // nothing
+#endif
+#ifndef OUT
+#    define OUT // nothing
+#endif
+#ifndef INOUT
+#    define INOUT // nothing
+#endif
+
+/**
+ * The type of PT trace's metadata. This type is used to initialize the PT config and PT
+ * sideband config of drpt2ir when decoding a PT trace.
+ */
+typedef struct __attribute__((__packed__)) _pt_metadata_t {
+    /** The CPU family. */
+    uint16_t cpu_family;
+    /** The CPU mode. */
+    uint8_t cpu_model;
+    /** The CPU stepping. */
+    uint8_t cpu_stepping;
+    /**
+     * The time shift. It is used to synchronize trace time, and the sideband recodes
+     * time.
+     * \note time_shift = perf_event_mmap_page.time_shift
+     */
+    uint16_t time_shift;
+    /**
+     * The time multiplier. It is used to synchronize trace time, and the sideband
+     * recodes time.
+     * \note time_mult = perf_event_mmap_page.time_mult
+     */
+    uint32_t time_mult;
+
+    /**
+     * The time zero. It is used to synchronize trace time, and the sideband recodes
+     * time.
+     * \note time_zero = perf_event_mmap_page.time_zero
+     */
+    uint64_t time_zero;
+} pt_metadata_t;
+
+/**
+ * The storage container type of a drpttracer's output.
+ * This data struct is used by drpttracer to store PT metadata, PT trace, and
+ * sideband data. These data can be dumped into different files. These files can be the
+ * inputs of drpt2ir, which decodes the PT data into Dynamorio's IR.
+ */
+typedef struct _drpttracer_buf_t {
+    pt_metadata_t metadata;   /**< The PT trace's metadata. */
+    void *pt_buf;             /**< The PT trace's buffer pointer. */
+    size_t pt_buf_size;       /**< The buffer size of PT trace. */
+    void *sideband_buf;       /**< The PT sideband data's buffer pointer. */
+    size_t sideband_buf_size; /**< The buffer size of PT sideband data. */
+} drpttracer_output_t;
+
+/***************************************************************************
+ * INIT
+ */
+
+DR_EXPORT
+/**
+ * Initializes the drpttracer extension.  Must be called prior to any of the
+ * other routines.  Can be called multiple times (by separate components,
+ * normally) but each call must be paired with a corresponding call to
```

Comment:
Do we really support calls by separate components? I believe (in at least the current implementation), only one component can use PT in one thread at a given time. Add to comments.

Reply:
Done.